File: 6) Go/tictactoe.go
package main

import "fmt"

func main() {

	board := []string{" ", " ", " ", " ", " ", " ", " ", " ", " "}

	fmt.Println("Welcome to tic-tac-toe!")

	xTurn(board)
}

func printBoard(board []string) {
	fmt.Println(board[0] + " | " + board[1] + " | " + board[2])
	fmt.Println("---------")
	fmt.Println(board[3] + " | " + board[4] + " | " + board[5])
	fmt.Println("---------")
	fmt.Println(board[6] + " | " + board[7] + " | " + board[8])
}

func xTurn(board []string) {
	printBoard(board)

	winCondition(board)

	fmt.Println("X, make your move! (1-9)")

	var spot int
	fmt.Scanln(&spot)

	// Check if spot is a number
	if !(spot > 0 && spot < 10) {
		fmt.Println("Invalid move!")
		xTurn(board)
	}

	// Check if spot is already taken
	if board[spot-1] != " " {
		fmt.Println("Spot is already taken!")
		xTurn(board)
	}

	board[spot-1] = "X"

	oTurn(board)
}

func oTurn(board []string) {
	printBoard(board)

	winCondition(board)

	fmt.Println("O, make your move! (1-9)")

	var spot int
	fmt.Scanln(&spot)

	// Check if spot is a number
	if !(spot > 0 && spot < 10) {
		fmt.Println("Invalid move!")
		oTurn(board)
	}

	// Check if spot is already taken
	if board[spot-1] != " " {
		fmt.Println("Spot is already taken!")
		oTurn(board)
	}

	board[spot-1] = "O"

	xTurn(board)
}

func winCondition(board []string) {
	resetBoard := []string{" ", " ", " ", " ", " ", " ", " ", " ", " "}
	switch checkWin(board) {
	case "X":
		fmt.Println("X wins!")
		xTurn(resetBoard)
	case "O":
		fmt.Println("O wins!")
		xTurn(resetBoard)
	case "tie":
		fmt.Println("Tie!")
		xTurn(resetBoard)
	}
}

func checkWin(board []string) string {
	// Winning combos
	combos := [][]int{
		{0, 1, 2},
		{3, 4, 5},
		{6, 7, 8},
		{0, 3, 6},
		{1, 4, 7},
		{2, 5, 8},
		{0, 4, 8},
		{2, 4, 6},
	}

	for _, combo := range combos {
		if board[combo[0]] == "X" && board[combo[1]] == "X" && board[combo[2]] == "X" {
			return "X"
		} else if board[combo[0]] == "O" && board[combo[1]] == "O" && board[combo[2]] == "O" {
			return "O"
		}
	}

	// Check for tie
	for _, spot := range board {
		if spot == " " {
			return "none"
		}
	}

	return "tie"
}

File: 7) Julia/TicTacToe.jl
function main()
    local board = [" ", " ", " ", " ", " ", " ", " ", " ", " "]

    println("Welcome to Tic Tac Toe!")

    xTurn(board)
end

function xTurn(board)
    printBoard(board)

    winCondition(board)

    print("X, make your move! (1-9) ")
    local move = readline()

    # Check if move is a number
    if (tryparse(Int64, move) !== nothing)
        move = tryparse(Int64, move)
        # Check if move is valid
        if (move > 0 && move < 10)
            # Check if move is empty
            if (board[move] == " ")
                board[move] = "X"
                oTurn(board)
            else
                println("That space is already taken!")
                xTurn(board)
            end
        else
            println("That is not a valid move!")
            xTurn(board)
        end
    else
        println("That is not a valid move!")
        xTurn(board)
    end
end

function oTurn(board)
    printBoard(board)

    winCondition(board)

    print("O, make your move! (1-9) ")
    local move = readline()

    # Check if move is a number
    if (tryparse(Int64, move) !== nothing)
        move = tryparse(Int64, move)
        # Check if move is valid
        if (move > 0 && move < 10)
            # Check if move is empty
            if (board[move] == " ")
                board[move] = "O"
                xTurn(board)
            else
                println("That space is already taken!")
                oTurn(board)
            end
        else
            println("That is not a valid move!")
            oTurn(board)
        end
    else
        println("That is not a valid move!")
        oTurn(board)
    end
end

function printBoard(board)
    println(" " * board[1] * " | " * board[2] * " | " * board[3])
    println("-----------")
    println(" " * board[4] * " | " * board[5] * " | " * board[6])
    println("-----------")
    println(" " * board[7] * " | " * board[8] * " | " * board[9])
end

function winCondition(board)
    local resetBoard = [" ", " ", " ", " ", " ", " ", " ", " ", " "]
    local win = checkWin(board)

    if (win == "X")
        println("X wins!")
        xTurn(resetBoard)
    elseif (win == "O")
        println("O wins!")
        xTurn(resetBoard)
    elseif (win == "Tie")
        println("Tie!")
        xTurn(resetBoard)
    else
        return
    end
end

function checkWin(board)
    local winConditions = [
        [1, 2, 3],
        [4, 5, 6],
        [7, 8, 9],
        [1, 4, 7],
        [2, 5, 8],
        [3, 6, 9],
        [1, 5, 9],
        [3, 5, 7]
    ]

    for winCondition in winConditions
        if (board[winCondition[1]] == "O" && board[winCondition[2]] == "O" && board[winCondition[3]] == "O")
            return "O"
        elseif (board[winCondition[1]] == "X" && board[winCondition[2]] == "X" && board[winCondition[3]] == "X")
            return "X"
        end
    end

    for i in board
        if(i == " ")
            return "None"
        end
    end

    return "Tie"
end

main()